### PR TITLE
Health bar battle feedback

### LINKED ISSUE #279: Read and Display Battle steps
Read the battle steps from the backend.
Assign a real time for each step to take.
In each step a feedback should be shown for the user to understand what is happening in the battle. To start a healthbar showing the live value of each unit while also displaying number indicators for the damage applied on hit.

### PR DESCRIPTION
**Pull request by tkz00**
Closes #279 
Waiting on backend PR [Integrate battles and levels](https://github.com/lambdaclass/mirra_backend/pull/468)

## Motivation

Feedback for unit health and damage has to be shown for the user to see this information if he chooses to simulate the battle instead of playing it out normally.

## Summary of changes

- A different protobuf message can be received from the backend on the Battle screen if the simulate button is selected on lineup, BattleReplay has all the information necessary to show battle feedback.
- The units are set up at the start of the battle and then the steps are played out, with health bars for the units that reflect the real value that they have and the damage taken for each unit is displayed with a fading number indicator.

## How has this been tested?

Go to the lineup of a level and choose the simulate button, the battle should play out with updating health bars and damage indicators.

### CODE REVIEW

**Comment by martinacantaro** [comment]
The simulation battle is 12 mufluses hitting each other at the same time. Every unit receives 50 damage each round, at the same time. But in the health bar I see some characters losing more health with each hit than others.

Let's assign numbers to each unit according to their position on screen, from top left to bottom right, like so:

1 --- 2 ---------- 3 --- 4

5 --- 6 ---------- 7 --- 8

9 --- 10 --------- 11 -- 12

I see unit 10 losing health much faster than all the others.
In step 3 it has lost all its health while most of the others have more than half left.
Then in step 4 unit 10 dies,  and unit 9 suddenly loses more than half of its total health in the bar. This seems to happen in two health drops during the same step, if you look closely at the video. At this step you can see a vast difference between the health of different units (screenshot below)

https://github.com/lambdaclass/champions_of_mirra/assets/6120897/baedbed8-e969-4e20-9f60-84dcc87ab1a3

Screenshot 2024-04-11 at 12 39 53

**Review by martinacantaro — CHANGES_REQUESTED** [review]
see comment above ☝️

**Comment by tkz00** [comment]
> The simulation battle is 12 mufluses hitting each other at the same time. Every unit receives 50 damage each round, at the same time. But in the health bar I see some characters losing more health with each hit than others.
> 
> Let's assign numbers to each unit according to their position on screen, from top left to bottom right, like so:
> 
> 1 --- 2 ---------- 3 --- 4
> 
> 5 --- 6 ---------- 7 --- 8
> 
> 9 --- 10 --------- 11 -- 12
> 
> I see unit 10 losing health much faster than all the others. In step 3 it has lost all its health while most of the others have more than half left. Then in step 4 unit 10 dies, and unit 9 suddenly loses more than half of its total health in the bar. This seems to happen in two health drops during the same step, if you look closely at the video. At this step you can see a vast difference between the health of different units (screenshot below)
> Screen.Recording.2024-04-11.at.12.31.49.mov
> Screenshot 2024-04-11 at 12 39 53

The targeting of the attack of each unit is random, that's why in some steps some units take more damage and others don't, and that random factor is what changes the outcome of the battle, winning some times and others loosing.

**Comment by martinacantaro** [comment]
Understood, so from the UX the text that shows the damage text appears different from what the health bar shows. All of them say -50 but some units have in fact multiple stacked "-50s", which appear as only one message because they are overlapping.

This is fine for now, but I will open a ticket to fix this after pre-Alpha

**Review by martinacantaro — DISMISSED** [review]

**Comment by martinacantaro** [comment]
per the comment above https://github.com/lambdaclass/champions_of_mirra/issues/347

**Review by martinacantaro — APPROVED** [review]
tested it again, still working fine!

File: client/Assets/Scripts/BackendConnection/SocketConnection.cs [hunk 1]
```diff
@@ -495,6 +493,89 @@ private void AwaitBattleResponse(byte[] data, Action<bool> onBattleResultReceive
...
+		WebSocketRequest request = new WebSocketRequest {
+			BattleTest = battleTestRequest
+		};
+		currentMessageHandler = (data) => AwaitFakeBattleResponse(data, onBattleReplayReceived);
+        ws.OnMessage += currentMessageHandler;
+        ws.OnMessage -= OnWebSocketMessage;
+        SendWebSocketMessage(request);
+		yield return null;
+
+		// // Create a new BattleReplay instance
```
Comment (ncontinanza) [inline client/Assets/Scripts/BackendConnection/SocketConnection.cs:528]: Is this commented code something that we should keep?

File: client/Assets/Scripts/Battle/BattleManager.cs [hunk 2]
```diff
@@ -14,31 +15,138 @@ public class BattleManager : MonoBehaviour
...
+					opponentUnitsIndex++;
+				}
+				units.Add(unit.Id, battleUnit);
+				battleUnit.gameObject.SetActive(true);
+				battleUnit.MaxHealth = unit.Health;
+				battleUnit.CurrentHealth = unit.Health;
+			}
+		#endregion
+
+		#region Play Out Battle
```
Comment (ncontinanza) [inline client/Assets/Scripts/Battle/BattleManager.cs:90]: Do you think we could break this into smaller methods to make it more readable and easier to understand?

File: gateway.proto [hunk 3]
```diff
@@ -151,9 +156,102 @@ option csharp_namespace = "Protobuf.Messages";
...
+    EFFECT_TRIGGER = 1;
+    EFFECT_HIT = 2;
+    EFFECT_MISS = 3;
+  }
+
+  enum Stat {
+    health = 0;
+    energy = 1;
+    damage = 2;
+    defense = 3;
```
Comment (ncontinanza) [inline gateway.proto:177]: These should be uppercase

**Review by ncontinanza — CHANGES_REQUESTED** [review]

File: client/Assets/Scripts/Battle/BattleManager.cs [hunk 4]
```diff
@@ -14,31 +15,138 @@ public class BattleManager : MonoBehaviour
...
+			battleReplay = replay;
+		}));
+
+		yield return new WaitUntil(() => battleReplay != null);
+
+		#region Set Up Initial State
+			int playerUnitsIndex = 0;
+			int opponentUnitsIndex = 0;
+			foreach (var unit in battleReplay.InitialState.Units)
+			// for(int i = 0; i < battleReplay.InitialState.Units.Count; i++)
```
Comment (ncontinanza) [inline client/Assets/Scripts/Battle/BattleManager.cs:65]: Is this commented code necessary?

**Review by ncontinanza — CHANGES_REQUESTED** [review]

File: client/Assets/Scripts/Battle/BattleManager.cs (hunk 4 above)
Comment (tkz00) [reply client/Assets/Scripts/Battle/BattleManager.cs]: Deleted!

File: gateway.proto (hunk 3 above)
Comment (tkz00) [reply gateway.proto]: Fixed!

File: client/Assets/Scripts/Battle/BattleManager.cs (hunk 2 above)
Comment (tkz00) [reply client/Assets/Scripts/Battle/BattleManager.cs]: You're right, those regions have been changed into separate methods!

File: client/Assets/Scripts/BackendConnection/SocketConnection.cs (hunk 1 above)
Comment (tkz00) [reply client/Assets/Scripts/BackendConnection/SocketConnection.cs]: Deleted

**Review by ncontinanza — APPROVED** [review]
LGTM!
I opened this issue to refactor the BattleManager later: https://github.com/lambdaclass/champions_of_mirra/issues/353

**Comment by tkz00** [comment]
> LGTM! I opened this issue to refactor the BattleManager later: #353

Great Nico! Just saw to comment! Many thanks!

**Review by marco-paulucci — APPROVED** [review]
LGTM, great work!

**Pull request merged**